File: components/src/main/java/org/nova/html/NumberFlow/number_flow.java
package org.nova.html.NumberFlow;

import org.nova.html.elements.TagElement;
import org.nova.html.ext.Content;
import org.nova.html.ext.HtmlUtils;
import org.nova.html.tags.script;
import org.nova.json.ObjectMapper;

public class number_flow extends TagElement<number_flow>
{
    public static final String SRC="https://cdn.jsdelivr.net/npm/number-flow@0.6.0/+esm";

    public static String js_set(String id,String value,Properties properties) throws Throwable
    {
        if (properties==null)
        {
            return HtmlUtils.js_call("barvian.number_flow.set", id, value);
        }
        else
        {
            return HtmlUtils.js_call("barvian.number_flow.set", id, value, ObjectMapper.writeObjectToString(properties));
        }
    }    
    public static String js_set(String id,String value) throws Throwable
    {
        return js_set(id,value,null);
    }    
    
    public number_flow()
    {
        super("number-flow");
    }
    
    public static class ScriptContent extends Content
    {
        public ScriptContent(String rootPath)
        {
            addInner(new script().type("module").src(SRC));
            addInner(new script().src(rootPath+"/barvian/number-flow.js"));
        }
    }
    
    public static enum Style
    {
        decimal("decimal"),
        percent("percent"),
        currency("currency"),
        unit("unit"),
        ;
        
        final private String value;

        private Style(String value) 
        {
            this.value = value;
        }

        @Override
        public String toString() 
        {
            return this.value;
        }        
    }
    public static class NumberFormatOptions
    {
        public Style style;
        public String currency;
    }
    
    public static class Properties
    {
        public NumberFormatOptions format;
        public String locales;
        public String numberPrefix;
        public Integer opacityTiming;
        public Integer transformTiming;
        public Integer spinTiming;
        public Integer trend;
        public Boolean animated;
        public Boolean respectMotionPreference;
    }
}

File: components/src/ts-0.1.4/barvian/number-flow.ts
namespace barvian.number_flow
{
    interface NumberFormatOptions
    {
        style:string;
        currency:string;
    }
    
    interface Properties
    {
        format: NumberFormatOptions;
        locales: String;
        numberPrefix: String;
        opacityTiming: Number;
        transformTiming: Number;
        spinTiming: Number;
        trend: Number;
        animated: Boolean;
        respectMotionPreference: Boolean;
    }

    export function set(id:string,value:string,properties:string=null):void
    {
        const flow = document.getElementById(id) as any;
        if (properties!==null)
        {
            const props=JSON.parse(properties) as Properties;
            if (props.format!==undefined)
            {
                flow.format={style: props.format.style, currency: props.format.currency};
            }
            if (props.locales!==undefined)
            {
                flow.locales=props.locales;
            }
            if (props.numberPrefix!==undefined)
            {
                flow.numberPrefix=props.numberPrefix;
            }
            if (props.opacityTiming!==undefined)
            {
                flow.opacityTiming=props.opacityTiming;
            }
            if (props.transformTiming!==undefined)
            {
                flow.transformTiming=props.transformTiming;
            }
            if (props.spinTiming!==undefined)
            {
                flow.spinTiming={duration: props.spinTiming};
            }
            if (props.trend!==undefined)
            {
                flow.trend=props.trend;
            }
            if (props.animated!==undefined)
            {
                flow.animated=props.animated;
            }
            if (props.respectMotionPreference!==undefined)
            {
                flow.respectMotionPreference=props.respectMotionPreference;
            }
        }
        flow.update(value);
        console.log("update: "+id+"="+value+",flow="+flow);
   }

}